src/proof/lfsc/lfsc_printer.cpp

```diff
@@ -44,7 +44,219 @@ void LfscPrinter::print(std::ostream& out,
                         const std::vector<Node>& assertions,
                         const ProofNode* pn)
 {
-  // to be added
+  Trace("lfsc-print-debug") << "; ORIGINAL PROOF: " << *pn << std::endl;
+  // closing parentheses
+  std::stringstream cparen;
+  const ProofNode* pnBody = pn->getChildren()[0].get();
+
+  // clear the rules we have warned about
+  d_trustWarned.clear();
+
+  Trace("lfsc-print-debug") << "; print declarations" << std::endl;
+  // [1] compute and print the declarations
+  std::unordered_set<Node> syms;
+  std::unordered_set<TNode> visited;
+  std::vector<Node> iasserts;
+  std::map<Node, size_t> passumeMap;
+  std::unordered_set<TypeNode> types;
+  std::unordered_set<TNode> typeVisited;
+  for (size_t i = 0, nasserts = assertions.size(); i < nasserts; i++)
+  {
+    Node a = assertions[i];
+    expr::getSymbols(a, syms, visited);
+    expr::getTypes(a, types, typeVisited);
+    iasserts.push_back(d_tproc.convert(a));
+    // remember the assumption name
+    passumeMap[a] = i;
+  }
+  d_assumpCounter = assertions.size();
+  Trace("lfsc-print-debug") << "; print sorts" << std::endl;
+  // [1a] user declared sorts
+  std::stringstream preamble;
+  std::unordered_set<TypeNode> sts;
+  std::unordered_set<size_t> tupleArity;
+  for (const TypeNode& st : types)
+  {
+    // note that we must get all "component types" of a type, so that
+    // e.g. U is printed as a sort declaration when we have type (Array U Int).
+    std::unordered_set<TypeNode> ctypes;
+    expr::getComponentTypes(st, ctypes);
+    for (const TypeNode& stc : ctypes)
+    {
+      if (sts.find(stc) != sts.end())
+      {
+        continue;
+      }
+      sts.insert(stc);
+      if (stc.isSort())
+      {
+        preamble << "(declare ";
+        printType(preamble, stc);
+        preamble << " sort)" << std::endl;
+      }
+      else if (stc.isDatatype())
+      {
+        const DType& dt = stc.getDType();
+        if (stc.getKind() == PARAMETRIC_DATATYPE)
+        {
+          // skip the instance of a parametric datatype
+          continue;
+        }
+        preamble << "; DATATYPE " << dt.getName() << std::endl;
+        if (dt.isTuple())
+        {
+          const DTypeConstructor& cons = dt[0];
+          size_t arity = cons.getNumArgs();
+          if (tupleArity.find(arity) == tupleArity.end())
+          {
+            tupleArity.insert(arity);
+            preamble << "(declare Tuple_" << arity << " ";
+            std::stringstream tcparen;
+            for (size_t j = 0, nargs = cons.getNumArgs(); j < nargs; j++)
+            {
+              preamble << "(! s" << j << " sort ";
+              tcparen << ")";
+            }
+            preamble << "sort" << tcparen.str() << ")";
+          }
+          preamble << std::endl;
+        }
+        else
```

Comment:
I think it would help to mention in a comment what are the remaining cases other than sorts and datatypes.